### ISSUE TITLE
I2I: amp-story-captions

### ISSUE THREAD

**Issue by hamax** [ampproject/amphtml#34016]
## Summary

To make captions more customizable and more consistent across browsers, we would like to introduce a new component that implements custom caption rendering, instead of relying on the native caption rendering.

## Motivation

Currently in AMP stories we can only use native captions rendering that comes with the video and track HTML tags. Format support and styling is largely browser dependent. Some styling workarounds are only possible with CSS !important, which is not allowed in AMP, but even with these workarounds the styling is limited.

### Style captions freely and consistently across browsers

Specify font size, margins and background.

Why this can’t be done with CSS: On Chrome, specifying font-size on ::cue is not enough, it also needs to be overwritten on ::-webkit-media-text-track-container using !important. On Safari, the background cannot be changed. Firefox doesn’t support styling to the same extent as WebKit browsers.

### Position captions relative to other elements on the page

Why this can’t be done with CSS: In CSS we can use transforms, which can move captions by some fixed amount. We can also modify the captions file to change the caption position. Both of these need to be specified before the page renders, so we don’t know the size of dynamic elements. It is possible to work around this with JS, but not with the existing AMP components.

### Support captions with timestamp tags (ASR/Karaoke style)

Based on my testing, [timestamp tags](https://www.w3.org/wiki/VTT_Concepts#Timestamp_Tags_.28Karaoke_Style_and_Paint_On_Caption_Text.29) within cues don’t seem to work in Firefox. We would like this style of captions to work across all browsers. We would also like to support custom styling for future and past parts of the cue.

### API

Example use within a story:

```
<amp-story-grid-layer>
  <amp-video id="video" src="..." captions-id="captions">
    <track src="..."></track>
  </amp-video>
</amp-story-grid-layer>
<amp-story-grid-layer>
  <amp-video-captions id="captions" layout="fixed-height" height="300"></amp-video-captions>
  <div>This text appears below captions</div>
</amp-story-grid-layer>
```

New component name: **amp-video-captions**

New attributes: **captions-id** on amp-video connects the two components

Layout: The layound needs to be a **size defined layout**. We don’t want the page to reflow every time a new caption cue is shown, and we don’t know how much space the captions need. (Could we support flex-item by setting flex-basis to 0?)

Styling:

The position of captions is controlled by the position of the amp-video-captions element. Properties like font-size and line-height can be specified using CSS on the amp-video-captions itself. To allow more granular control, we expose additional CSS class **amp-video-captions-future** that control the style of future parts of the cue for karaoke-style captions. We might need to introduce additional CSS classes for more granular styling.

Example CSS:

```
amp-video-captions {
    color: white;
    font-size: 24px;
    padding: 16px;
}

// Words not spoken yet shown in gray.
.amp-video-captions-future {
    color: gray;
}
```

#### amp-video and amp-video-captions communication

When &lt;video> is initialized or reset in amp-video (resetOnDomChange called), amp-video calls amp-video-captions and passes it a reference to the video element. The captions component linstents to any changes on the textTracks object. Adding, removing, or making tracks shown/hidden triggers a textTracks change event.

#### Hiding native caption rendering

Native caption rendering can be hidden by setting track.mode to hidden. To simplify the implementation, amp-video-captions sets every track with mode showing to hidden, and renders all tracks that are set to hidden.

There are two potential issues with this:

*   Tracks that were meant to be hidden will be rendered. If there’s a need for hidden tracks outside of this component, we could keep track of which tracks were set to hidden by the component itself.
*   Since the component is in a separate extension to amp-video, the track might be set to shown for a while before the component is loaded. If this is an issue, the “hiding” logic can be moved to amp-video (because amp-video knows when there’s an amp-video-captions component attached to it).

#### Rendering

Each track is rendered by a TrackRenderer. In the vast majority of cases, there’s only one track visible at one time, but we do support multiple visible tracks. The TrackRenderer needs to listen to two events:

*   Cuechange: This event triggers when the active cues change. When this happens, TrackRenderer converts all active cues to HTML elements.
*   Timeupdate: This event triggers as video plays or seeks to a different position. For karaoke-style captions, TrackRenderer needs to update which parts of the cues are in the past/future.

To minimize the work that needs to be done on the frequent timeupdate events, the cues are split into sections around the [timestamp tags](https://www.w3.org/wiki/VTT_Concepts#Timestamp_Tags_.28Karaoke_Style_and_Paint_On_Caption_Text.29). This way the timeupdate event handler only needs to update the amp-video-captions-future class on the section elements instead of rerendering everything. Note that there is no event that would trigger when a timestamp tag boundary is crossed (the equivalent to cuechange), so we need to listen to timeupdate events.

#### Other WebVTT features

Each cue can have an alignment and position. These are used to position the cue element within the component’s bounds (e.g. left/center/right aligned, top/bottom positioned). There are some other features listed [here](https://www.w3.org/wiki/VTT_Concepts#Optional_Cue_Settings) that are not yet implemented and will need to be added for full feature parity with the native rendering.

### Better support for styling native captions

Amp-video could accept optional caption positioning parameters and dynamically move captions to that position. This parameter could accept id of another element (e.g. “above #some-div”). Using this, the amp-video JS could reposition the captions using either CSS or caption line property.

This solution only solves positioning issues, and not other issues mentioned above.

### Implementing a custom WebVTT parser

This doesn’t seem to be needed currently, and it would significantly increase the code size. If we identify features that would require this in the future, it’s possible to do without changing the API.

## Implementation in progress

https://github.com/ampproject/amphtml/compare/main...hamax:captions

**Comment by newmuis** [comment]
This is fantastic, and will provide so much more flexibility in terms of appearance of captions.

I know there is also desire to use this for audio components (for voiceovers and speech to text), so maybe we take the `-video-` infix out and call it `amp-captions`?

**Comment by alanorozco** [comment]
This looks like a good idea. Has this been registered to go through [Design Review](https://github.com/ampproject/amphtml/issues?q=is%3Aissue+is%3Aopen+label%3A%22Type%3A+Design+Review%22)? 

Also, one suggestion. Instead of:

```html
  <amp-video id="video" src="..." captions-id="captions">
    <track src="..."></track>
  </amp-video>
  <amp-video-captions id="captions" layout="fixed-height" height="300"></amp-video-captions>
```

I would suggest that the captions element wraps the element instead, to make sure that captions are displayed on top of the element ('z-index, pointer-events: none, etc.'). You also forgo the need to cross-reference by id:


```html
  <amp-captions layout="responsive" width="16" height="9">
    <amp-video layout="fill" ...>
    </amp-video>
  </amp-captions>
```

**Comment by newmuis** [comment]
Good idea to bring it to design review.  @hamax and @gmajoulet, can you make that happen?

> I would suggest that the captions element wraps the element instead

This actually might make certain layouts harder to achieve, given the layer system in stories.  It's quite possible that you don't actually want the captions and the video to be positioned similarly (and is, in fact, one of the things I'm excited to have solved by this new syntax).

**Comment by alanorozco** [comment]
@newmuis Could we make sure that we present what we want this to look like? An HTML interface that would suit both Stories and standard Documents would be ideal, so we might need a mockup of what's wanted on Stories to understand that better.

**Comment by hamax** [comment]
Like @newmuis said, amp-video and amp-video-captions being positioned independently is one of the features of this new component (the main motivation). I agree that it would be good to make it easy to use for the simple use case as well, where we don't need to reposition the captions. That could be the API that @alanorozco proposed, or even just amp-video automatically using this component instead of the native captions by default or enabled by an attribute.

**Comment by hamax** [comment]
In the design review we discussed the additional issues this would have when used outside of stories, and decided it would be best to make this story only for now. For posterity, one of the issues is alignment with native video controls. Combined with the suggestion from @newmuis above, I propose we change the name to *amp-story-captions*.

Another thing brought up was the name of the *captions-id* attribute. The *label* HTML tag uses *for* attibute to refer to form elements by id, so we could follow that. This would reverse the direction (captions would refer to video), which might not be desirable, so I'm not sure if it's better. The downside of the reverse direction is that the video component doesn't know that the captions are being rendered externally until it gets a call from the captions component. In the current implementation this information is not used, but it could be useful in the future.

**Comment by juankk** [comment]
Newscorp Australia needs this AMP Video Caption fix to unblock us with adding captions. At the moment the Editorial team needs to add the captions in video before uploading to the WordPress Media Library for web stories. The back and forth to get captions into the video and then make sure those captions are in the Web Story safe area is adding 60 to 90 minutes to Web Story builds requiring video. The lack of captions in the Web Story also lowers the amount of text in the web story, impacting discoverability through search.

Seems that there is an open PR pending to have merge conflicts resolved. Is there a way of knowing when will this change be available? @gmajoulet

**Comment by gmajoulet** [comment]
The code was just merged, and will be able in beta next Tuesday, and in production the following Tuesday.

**Comment by jingfei** [comment]
Thanks @gmajoulet, this is exactly the feature we're looking to, very excited for the new component!

**Comment by gmajoulet** [comment]
I'd recommend using an `amp-video` component that's loading your audio source, and no image content. You'll be able to configure and style the `amp-story-captions`. Feel free to open a new issue and cc us on it if should you have any problem :))

**Comment by jingfei** [comment]
@gmajoulet - Sure, we'll try the approach. Also, we'd like to know if there's any easy way (e.g. a js script to enable the experiment) for us to test the feature before it's 100% released to prod? Thank you.

**Comment by gmajoulet** [comment]
You can [opt-in the beta channel here](https://cdn.ampproject.org/experiments.html), it's a per-browser opt-in though. Otherwise the code will get to production at some point next Tuesday!

**Comment by jingfei** [comment]
Hi @gmajoulet we found that the component supports only the layout that depends on the parent element or needs a predefined width/height, and we're wondering if you could help to enable `container` layout for the caption to define the element size by content? Thank you.

**Comment by gmajoulet** [comment]
Yes that sounds fine, do you think you could help with this change?

You'd need to update
https://github.com/ampproject/amphtml/blob/2d5d9a295d45fdfe1abe4079a396c72a1dcbc328/extensions/amp-story-captions/0.1/amp-story-captions.js#L35-L38

Into something like:
```
import {Layout, isLayoutSizeDefined} from '#core/dom/layout';

...

return layout == Layout.CONTAINER || isLayoutSizeDefined(layout);
```

And then enabling validation here:

https://github.com/ampproject/amphtml/blob/2d5d9a295d45fdfe1abe4079a396c72a1dcbc328/extensions/amp-story-captions/validator-amp-story-captions.protoascii#L34-L42

**Comment by jingfei** [comment]
Sure, thanks for the pointer. I have a question regarding the layout -
can we render the element differently base on layout like `if (layout === CONTAINER) { /** container layout and style settings */ }` because:

1. The main element of amp-story-captions invoke [`applyFillContent`](https://github.com/ampproject/amphtml/blob/6283538b66cbf0e2bf404d7b81a02ef2b2715507/src/core/dom/layout/index.js#L247) to use "fill content" layout, and that makes the element to depend on parent's size
https://github.com/ampproject/amphtml/blob/55fa3e004e2aeece5faf9a68717228077c215a13/extensions/amp-story-captions/0.1/amp-story-captions.js#L32
2. Similarly the child element of amp-story-captions uses styles that make the content to depend on parent's size
https://github.com/ampproject/amphtml/blob/55fa3e004e2aeece5faf9a68717228077c215a13/extensions/amp-story-captions/0.1/track-renderer.js#L77-L83

It'd help supporting amp-story-captions container layout if we can skip `applyFillContent` call and the absolute style settings for the certain layout.

**Comment by gmajoulet** [comment]
That sounds good to me, cc @alanorozco to double check

**Comment by jingfei** [comment]
Thanks, it'd be great if you can guide me about the way to get `layout` variable from `buildCallback` and `setVideoElement` method?

**Comment by jingfei** [comment]
I'd like to follow-up on this~
@alanorozco could you help to take a look? Thanks.

**Comment by alanorozco** [comment]
Sounds like the plan works.

**Comment by pietrop** [comment]
Are there any docs for the final version of this?
The docs seems to only point to this issue https://amp.dev/documentation/components/amp-story-captions
